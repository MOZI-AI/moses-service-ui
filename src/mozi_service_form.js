import React from 'react';
import { MosesOptionsForm } from './moses_opts';
import { CrossValidationOptionsForm } from './crossval_opts';
import { DatasetUpload } from './dataset_upload';
import { stringifyMosesOptions } from './utils';
import { TargetFeatureForm } from './target_feature';
import {
  AnalysisParameters,
  CrossValOptions,
  Filter
} from './proto/moses_service_pb';
import { Stepper, Step, StepLabel, Divider } from '@material-ui/core';

const Options = {
  DATASET: 0,
  MOSES_OPTIONS: 1,
  CROSS_VALIDATION_OPTIONS: 2,
  TARGET_FEATURE: 3
};

export class MoziServiceForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      mosesOpts: {
        maximumEvals: 1000,
        featureSelectionTargetSize: 4,
        reductKnobBuildingEffort: '0',
        resultCount: 100,
        numberOfThreads: 8,
        featureSelectionAlgorithm: 'simple',
        enableFeatureSelection: true,
        hcWidenSearch: true,
        balance: true,
        hcCrossoverMinNeighbors: 5000,
        hcCrossoverPopSize: 1000,
        complexityRatio: 3
      },
      additionalParameters: [],
      crossValOptions: {
        folds: 1,
        testSize: 0.3,
        randomSeed: 5
      },
      dataset: undefined,
      targetFeature: 'case',
      filter: { name: '', value: '' },
      datasetFile: undefined,
      currentStep: Options.DATASET
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleAdditionalParametersAdded = this.handleAdditionalParametersAdded.bind(
      this
    );
    this.handleAdditionalParametersRemoved = this.handleAdditionalParametersRemoved.bind(
      this
    );
    this.handleFileUpload = this.handleFileUpload.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.changeStep = this.changeStep.bind(this);
    this.handleFilterChange = this.handleFilterChange.bind(this);
  }

  changeStep(step) {
    this.setState({ currentStep: step });
  }

  handleFileUpload(file) {
    const fileReader = new FileReader();
    fileReader.readAsDataURL(file);
    fileReader.onload = () => {
      let encoded = fileReader.result.replace(/^data:(.*;base64,)?/, '');
      encoded.length % 4 > 0 &&
        (encoded += '='.repeat(4 - (encoded.length % 4)));
      this.setState({ dataset: encoded, datasetFile: file });
    };
  }

  handleInputChange(option, event) {
    const target = event.target;
    const update = { [target.name]: this.parseValue(target) };
    this.setState(state => {
      switch (option) {
        case Options.MOSES_OPTIONS:
          return { mosesOpts: Object.assign({}, state.mosesOpts, update) };
        case Options.CROSS_VALIDATION_OPTIONS:
          return {
            crossValOptions: Object.assign({}, state.crossValOptions, update)
          };
        case Options.TARGET_FEATURE:
          return { targetFeature: update[target.name] };
        default:
          break;
      }
    });
    event.preventDefault();
  }

  handleFilterChange(update) {
    this.setState(state => ({
      filter: Object.assign({}, state.filter, update)
    }));
  }

  handleAdditionalParametersAdded(name, value) {
    const param = Object.assign({}, this.state.additionalParameters, {
      [name]: value
    });
    this.setState({ additionalParameters: param });
  }

  handleAdditionalParametersRemoved(name) {
    const additionalParameters = Object.assign(
      {},
      this.state.additionalParameters
    );
    delete additionalParameters[name];
    this.setState({ additionalParameters: additionalParameters });
  }

  parseValue(target) {
    if (target.value === '') return '';
    let parsedValue =
      target.type === 'checkbox'
        ? !!+target.checked
        : isNaN(target.value)
        ? target.value
        : +target.value;
    return parsedValue;
  }

  handleSubmit() {
    // if a request is being processed, do nothing and wait for it to finish
    if (this.props.busy) {
      return;
    }
    const crossValOptions = new CrossValOptions();
    crossValOptions.setFolds(this.state.crossValOptions.folds);
    crossValOptions.setRandomseed(this.state.crossValOptions.randomSeed);
    crossValOptions.setTestsize(this.state.crossValOptions.testSize);

    const analysisParameters = new AnalysisParameters();
    analysisParameters.setMosesopts(
      stringifyMosesOptions(
        this.state.mosesOpts,
        this.state.additionalParameters
      )
    );

    const filter = new Filter();
    filter.setScore(this.state.filter.name);
    filter.setValue(this.state.filter.name ? this.state.filter.value : -1);

    analysisParameters.setDataset(this.state.dataset);
    analysisParameters.setTargetFeature(this.state.targetFeature);
    analysisParameters.setCrossvalopts(crossValOptions);
    analysisParameters.setFilter(filter);

    this.props.handleSubmit(analysisParameters);
  }

  render() {
    return (
      <React.Fragment>
        <Stepper activeStep={this.state.currentStep} style={{ color: 'red' }}>
          <Step key={1}>
            <StepLabel>Select dataset</StepLabel>
          </Step>
          <Step key={1}>
            <StepLabel>Moses Options</StepLabel>
          </Step>
          <Step key={1}>
            <StepLabel>Cross Validation Options</StepLabel>
          </Step>
          <Step key={1}>
            <StepLabel>Target feature</StepLabel>
          </Step>
        </Stepper>

        {this.props.error && (
          <div
            style={{
              padding: '15px',
              border: 'solid 1px #FFA19D',
              backgroundColor: '#FFF0EF',
              borderRadius: '5px'
            }}
          >
            <p>{this.props.error}</p>
          </div>
        )}
        <Divider style={{ marginBottom: '30px' }} />
        <DatasetUpload
          show={this.state.currentStep === Options.DATASET}
          uploadedFile={this.state.datasetFile}
          handleFileUpload={this.handleFileUpload}
          next={() => this.changeStep(Options.MOSES_OPTIONS)}
        />
        <MosesOptionsForm
          show={this.state.currentStep === Options.MOSES_OPTIONS}
          defaults={this.state.mosesOpts}
          additionalParameters={this.state.additionalParameters}
          changeInput={event =>
            this.handleInputChange(Options.MOSES_OPTIONS, event)
          }
          addAdditionalParameter={this.handleAdditionalParametersAdded}
          removeAdditionalParameter={this.handleAdditionalParametersRemoved}
          back={() => this.changeStep(Options.DATASET)}
          next={() => this.changeStep(Options.CROSS_VALIDATION_OPTIONS)}
        />
        <CrossValidationOptionsForm
          show={this.state.currentStep === Options.CROSS_VALIDATION_OPTIONS}
          defaults={this.state.crossValOptions}
          changeInput={event =>
            this.handleInputChange(Options.CROSS_VALIDATION_OPTIONS, event)
          }
          back={() => this.changeStep(Options.MOSES_OPTIONS)}
          next={() => this.changeStep(Options.TARGET_FEATURE)}
        />
        <TargetFeatureForm
          show={this.state.currentStep === Options.TARGET_FEATURE}
          defaults={{
            targetFeature: this.state.targetFeature,
            filter: this.state.filter
          }}
          changeInput={event =>
            this.handleInputChange(Options.TARGET_FEATURE, event)
          }
          back={() => this.changeStep(Options.CROSS_VALIDATION_OPTIONS)}
          submit={this.handleSubmit}
          handleFilterChange={this.handleFilterChange}
        />
      </React.Fragment>
    );
  }
}
