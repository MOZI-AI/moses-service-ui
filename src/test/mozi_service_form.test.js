import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { MoziServiceForm } from '../mozi_service_form';

describe('<MoziServiceForm />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<MoziServiceForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows dataset upload form on start', () => {
    const wrapper = shallow(
      <MoziServiceForm handleSubmit={() => {}} show="true" />
    );
    expect(wrapper.find('DatasetUpload').prop('show')).toBeTruthy();
  });
});
