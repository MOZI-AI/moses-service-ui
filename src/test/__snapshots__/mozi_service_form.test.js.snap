// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoziServiceForm /> renders correctly 1`] = `
Array [
  <div
    className="MuiPaper-root-5 MuiPaper-elevation0-7 MuiStepper-root-1 MuiStepper-horizontal-2"
    style={
      Object {
        "color": "red",
      }
    }
  >
    <div
      className="MuiStep-root-32 MuiStep-horizontal-33"
    >
      <span
        className="MuiStepLabel-root-37 MuiStepLabel-horizontal-38"
      >
        <span
          className="MuiStepLabel-iconContainer-45"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-53 MuiStepIcon-root-48 MuiStepIcon-active-50"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="12"
            />
            <text
              className="MuiStepIcon-text-49"
              textAnchor="middle"
              x="12"
              y="16"
            >
              1
            </text>
          </svg>
        </span>
        <span
          className="MuiStepLabel-labelContainer-47"
        >
          <span
            className="MuiTypography-root-62 MuiTypography-body1-71 MuiStepLabel-label-40 MuiStepLabel-active-41"
          >
            Select dataset
          </span>
        </span>
      </span>
    </div>
    <div
      className="MuiStepConnector-root-98 MuiStepConnector-horizontal-99 MuiStepConnector-disabled-104"
    >
      <span
        className="MuiStepConnector-line-105 MuiStepConnector-lineHorizontal-106"
      />
    </div>
    <div
      className="MuiStep-root-32 MuiStep-horizontal-33"
    >
      <span
        className="MuiStepLabel-root-37 MuiStepLabel-horizontal-38 MuiStepLabel-disabled-44"
      >
        <span
          className="MuiStepLabel-iconContainer-45"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-53 MuiStepIcon-root-48"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="12"
            />
            <text
              className="MuiStepIcon-text-49"
              textAnchor="middle"
              x="12"
              y="16"
            >
              2
            </text>
          </svg>
        </span>
        <span
          className="MuiStepLabel-labelContainer-47"
        >
          <span
            className="MuiTypography-root-62 MuiTypography-body1-71 MuiStepLabel-label-40"
          >
            Moses Options
          </span>
        </span>
      </span>
    </div>
    <div
      className="MuiStepConnector-root-98 MuiStepConnector-horizontal-99 MuiStepConnector-disabled-104"
    >
      <span
        className="MuiStepConnector-line-105 MuiStepConnector-lineHorizontal-106"
      />
    </div>
    <div
      className="MuiStep-root-32 MuiStep-horizontal-33"
    >
      <span
        className="MuiStepLabel-root-37 MuiStepLabel-horizontal-38 MuiStepLabel-disabled-44"
      >
        <span
          className="MuiStepLabel-iconContainer-45"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-53 MuiStepIcon-root-48"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="12"
            />
            <text
              className="MuiStepIcon-text-49"
              textAnchor="middle"
              x="12"
              y="16"
            >
              3
            </text>
          </svg>
        </span>
        <span
          className="MuiStepLabel-labelContainer-47"
        >
          <span
            className="MuiTypography-root-62 MuiTypography-body1-71 MuiStepLabel-label-40"
          >
            Cross Validation Options
          </span>
        </span>
      </span>
    </div>
    <div
      className="MuiStepConnector-root-98 MuiStepConnector-horizontal-99 MuiStepConnector-disabled-104"
    >
      <span
        className="MuiStepConnector-line-105 MuiStepConnector-lineHorizontal-106"
      />
    </div>
    <div
      className="MuiStep-root-32 MuiStep-horizontal-33"
    >
      <span
        className="MuiStepLabel-root-37 MuiStepLabel-horizontal-38 MuiStepLabel-disabled-44"
      >
        <span
          className="MuiStepLabel-iconContainer-45"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-53 MuiStepIcon-root-48"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="12"
            />
            <text
              className="MuiStepIcon-text-49"
              textAnchor="middle"
              x="12"
              y="16"
            >
              4
            </text>
          </svg>
        </span>
        <span
          className="MuiStepLabel-labelContainer-47"
        >
          <span
            className="MuiTypography-root-62 MuiTypography-body1-71 MuiStepLabel-label-40"
          >
            Target feature
          </span>
        </span>
      </span>
    </div>
  </div>,
  <hr
    className="MuiDivider-root-108"
    style={
      Object {
        "marginBottom": "30px",
      }
    }
  />,
  <div
    className="datasetUploadWrapper"
  >
    <div
      className="dropzone"
      onBlur={[Function]}
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      style={
        Object {
          "border": "dashed 1px #90D4FF",
          "padding": "30px",
          "textAlign": "center",
        }
      }
      tabIndex={0}
    >
      <input
        autoComplete="off"
        multiple={false}
        name="dataset"
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        tabIndex={-1}
        type="file"
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-53"
        focusable="false"
        role="presentation"
        style={
          Object {
            "fontSize": "48px",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
        />
      </svg>
      <p>
        Click here to select a dataset file, or drag and drop it over this text.
      </p>
    </div>
    <div
      className="MuiGrid-container-113 MuiGrid-spacing-xs-24-137"
      style={
        Object {
          "justifyItems": "end",
          "marginTop": "30px",
        }
      }
    >
      <div
        className="MuiGrid-item-114 MuiGrid-grid-xs-12-153"
      >
        <button
          className="MuiButtonBase-root-236 MuiButtonBase-disabled-237 MuiButton-root-210 MuiButton-contained-221 MuiButton-containedPrimary-222 MuiButton-raised-224 MuiButton-raisedPrimary-225 MuiButton-disabled-230"
          disabled={true}
          id="nextButton"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="-1"
          type="button"
        >
          <span
            className="MuiButton-label-211"
          >
            Next
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-53"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
